lib/migration/config/v1/virtual_container_host.go

```diff
@@ -0,0 +1,317 @@
+	t.ExecutorConfig.Key = *key
+}
+
+// SetName sets the name of the VCH - this will be used as the hostname for the appliance
+func (t *VirtualContainerHostConfigSpec) SetName(name string) {
+	t.ExecutorConfig.Name = name
+}
+
+// SetDebug configures the debug logging level for the VCH
+func (t *VirtualContainerHostConfigSpec) SetDebug(level int) {
+	t.ExecutorConfig.Diagnostics.DebugLevel = level
+}
+
+// SetMoref sets the moref of the VCH - this allows components to acquire a handle to
+// the appliance VM.
+func (t *VirtualContainerHostConfigSpec) SetMoref(moref *types.ManagedObjectReference) {
+	if moref != nil {
+		t.ExecutorConfig.ID = moref.String()
+	}
+}
+
+// AddNetwork adds a network that will be configured on the appliance VM
+func (t *VirtualContainerHostConfigSpec) AddNetwork(net *executor.NetworkEndpoint) {
+	if net != nil {
+		if t.ExecutorConfig.Networks == nil {
+			t.ExecutorConfig.Networks = make(map[string]*executor.NetworkEndpoint)
+		}
+
+		t.ExecutorConfig.Networks[net.Network.Name] = net
+	}
+}
+
+// AddContainerNetwork adds a network that will be configured on the appliance VM
+func (t *VirtualContainerHostConfigSpec) AddContainerNetwork(net *executor.ContainerNetwork) {
+	if net != nil {
+		if t.ContainerNetworks == nil {
+			t.ContainerNetworks = make(map[string]*executor.ContainerNetwork)
+		}
+
+		t.ContainerNetworks[net.Name] = net
+	}
+}
+
+func (t *VirtualContainerHostConfigSpec) AddComponent(name string, component *executor.SessionConfig) {
+	if component != nil {
+		if t.ExecutorConfig.Sessions == nil {
+			t.ExecutorConfig.Sessions = make(map[string]*executor.SessionConfig)
+		}
+
+		if component.Name == "" {
+			component.Name = name
+		}
+		if component.ID == "" {
+			component.ID = name
+		}
+		t.ExecutorConfig.Sessions[name] = component
+	}
+}
+
+func (t *VirtualContainerHostConfigSpec) AddImageStore(url *url.URL) {
+	if url != nil {
+		t.ImageStores = append(t.ImageStores, *url)
+	}
+}
+
+func (t *VirtualContainerHostConfigSpec) AddVolumeLocation(name string, u *url.URL) {
+
+	if u != nil {
+		if t.VolumeLocations == nil {
+			t.VolumeLocations = make(map[string]*url.URL)
+		}
+
+		t.VolumeLocations[name] = u
+	}
+}
+
+// AddComputeResource adds a moref to the set of permitted root pools. It takes a ResourcePool rather than
+// an inventory path to encourage validation.
+func (t *VirtualContainerHostConfigSpec) AddComputeResource(pool *types.ManagedObjectReference) {
+	if pool != nil {
+		t.ComputeResources = append(t.ComputeResources, *pool)
+	}
+}
+
+func CreateSession(cmd string, args ...string) *executor.SessionConfig {
+	cfg := &executor.SessionConfig{
+		Cmd: executor.Cmd{
+			Path: cmd,
+			Args: []string{
+				cmd,
+			},
+		},
+	}
+
+	cfg.Cmd.Args = append(cfg.Cmd.Args, args...)
+
+	return cfg
+}
+
+func (t *RawCertificate) Certificate() (*tls.Certificate, error) {
+	if t.IsNil() {
+		return nil, errors.New("nil certificate")
```

Comment:
I would make this error as a part of API. Something like:

```
var NilCertErr = errors.New("nil certificate") 

func (t *RawCertificate) X509Certificate() (*x509.Certificate, error) {
 	if t.IsNil() {
 		return nil, NilCertErr
 	}
 	cert, _, err := certificate.ParseCertificate(t.Cert, t.Key)
 	return cert, err
 }

```



So, I could clearly check if the certificate is actually nil avoiding string comparison.